templates/usr/local/sbin/simp_generate_types.epp

```diff
@@ -0,0 +1,104 @@
+#!/opt/puppetlabs/puppet/bin/ruby
+
+require 'ostruct'
+require 'optparse'
+require 'syslog/logger'
+
+log = Syslog::Logger.new(File.basename(__FILE__))
+
+puppet_cmd = '/opt/puppetlabs/bin/puppet'
+
+environment_paths = Dir.glob(
+  %x{#{puppet_cmd} config print --section=master environmentpath}.strip.split(':').map{|x| x + '/*'}
+)
+
+options = OpenStruct.new
+options.quiet = false
+options.path = nil
+options.syslog = false
+options.target_environments = []
+
+OptionParser.new do |opts|
+  opts.on('-p', '--path TGT_PATH') do |path|
+    options.path = path
+
+    if path == 'all'
+      options.target_environments = environment_paths.compact
+    else
+      options.target_environments = Array(environment_paths.find do |env|
+        path.start_with?(env)
+      end).compact
+    end
+
+    options.target_environments.map!{|e| File.basename(e)}
+  end
+  opts.on('-l', '--logfile LOGFILE') do |file|
+    options.logfile = file
+  end
+  opts.on('-s', '--syslog') do
+    options.syslog = true
+  end
+  opts.on('-q', '--quiet') do
+    options.quiet = true
+  end
+  opts.on('-h', '--help', 'This help message') do
+    puts opts
+    exit
+  end
+
+  options.help = opts.help
+
+  end.parse!
+
+fh = File.open(options.logfile, 'w') if options.logfile
+
+unless options.path
+  msg = "Error: You must pass a path"
+
+  $stderr.puts msg
+  $stderr.puts options.help
+
+  log.error(msg) if options.syslog
```

Comment:
this repeated log code lends itself to a log method that does the correct thing, if the body of the script was encapsulated into a class that stored the options in an instance variable

Reply:
Will keep this in mind for round 2